libbeat/esleg/eslegclient/connection.go

```diff
@@ -0,0 +1,435 @@
+
+	dialer = transport.NetDialer(s.Timeout)
+	tlsDialer, err = transport.TLSDialer(dialer, s.TLS, s.Timeout)
+	if err != nil {
+		return nil, err
+	}
+
+	if st := s.Observer; st != nil {
+		dialer = transport.StatsDialer(dialer, st)
+		tlsDialer = transport.StatsDialer(tlsDialer, st)
+	}
+
+	var encoder BodyEncoder
+	compression := s.CompressionLevel
+	if compression == 0 {
+		encoder = NewJSONEncoder(nil, s.EscapeHTML)
+	} else {
+		encoder, err = NewGzipEncoder(compression, nil, s.EscapeHTML)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var proxy func(*http.Request) (*url.URL, error)
+	if !s.ProxyDisable {
+		proxy = http.ProxyFromEnvironment
+		if s.Proxy != nil {
+			proxy = http.ProxyURL(s.Proxy)
+		}
+	}
+
+	return &Connection{
+		ConnectionSettings: s,
+		HTTP: &http.Client{
+			Transport: &http.Transport{
+				Dial:            dialer.Dial,
+				DialTLS:         tlsDialer.Dial,
+				TLSClientConfig: s.TLS.ToConfig(),
+				Proxy:           proxy,
+			},
+			Timeout: s.Timeout,
+		},
+		Encoder: encoder,
+		log:     logp.NewLogger("esclientleg"),
+	}, nil
+}
+
+// NewClients returns a list of Elasticsearch clients based on the given
+// configuration. It accepts the same configuration parameters as the Elasticsearch
+// output, except for the output specific configuration options.  If multiple hosts
+// are defined in the configuration, a client is returned for each of them.
+func NewClients(cfg *common.Config) ([]Connection, error) {
+	config := defaultConfig
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, err
+	}
+
+	tlsConfig, err := tlscommon.LoadTLSConfig(config.TLS)
+	if err != nil {
+		return nil, err
+	}
+
+	var proxyURL *url.URL
+	if !config.ProxyDisable {
+		proxyURL, err = common.ParseURL(config.ProxyURL)
+		if err != nil {
+			return nil, err
+		}
+		if proxyURL != nil {
+			logp.Info("using proxy URL: %s", proxyURL)
+		}
+	}
+
+	params := config.Params
+	if len(params) == 0 {
+		params = nil
+	}
+
+	clients := []Connection{}
+	for _, host := range config.Hosts {
+		esURL, err := common.MakeURL(config.Protocol, config.Path, host, 9200)
+		if err != nil {
+			logp.Err("invalid host param set: %s, Error: %v", host, err)
+			return nil, err
+		}
+
+		client, err := NewConnection(ConnectionSettings{
+			URL:              esURL,
+			Proxy:            proxyURL,
+			ProxyDisable:     config.ProxyDisable,
+			TLS:              tlsConfig,
+			Username:         config.Username,
+			Password:         config.Password,
+			APIKey:           config.APIKey,
+			Parameters:       params,
+			Headers:          config.Headers,
+			Timeout:          config.Timeout,
+			CompressionLevel: config.CompressionLevel,
+		})
+		if err != nil {
+			return clients, err
+		}
+		clients = append(clients, *client)
+	}
+	if len(clients) == 0 {
+		return clients, fmt.Errorf("no hosts defined in the config")
+	}
+	return clients, nil
+}
+
+func NewConnectedClient(cfg *common.Config) (*Connection, error) {
```

Comment:
Where do we need this? It would be nice if we would not have to pass `*common.Config`, but a concreate configuration object instead.

Reply:
This is being called from the following places:

* `filebeat/beater/filebeat.go`: https://github.com/elastic/beats/pull/16150/files#diff-4a01e68d6f9559f53bacbaa546d541d2R153
* `filebeat/beater/filebeat.go`: https://github.com/elastic/beats/pull/16150/files#diff-4a01e68d6f9559f53bacbaa546d541d2R361
* `libbeat/cmd/instance/beat.go`: https://github.com/elastic/beats/pull/16150/files#diff-f7226280e7734d6e3d88e7dde232602eR501

In each case the caller is passing the Elasticsearch output config to this function.

I've been thinking of providing a helper/utility function that would take an Elasticsearch output config (`common.Config`) and convert it to a `eslegclient.ConnectionSettings` struct. Then code like in the three places above can use this utility. WDYT?

Reply:
We aim to remove this package long term. Maybe not worth the effort.